File: src/ai_node/bootstrap/bootstrap_parser.py
import json
from typing import Iterable, Tuple
from urllib.parse import urljoin, urlparse

from ai_node.security.boundaries import enforce_bootstrap_security_boundary


REQUIRED_FIELDS = (
    "topic",
    "bootstrap_version",
    "core_id",
    "core_name",
    "core_version",
    "api_base",
    "mqtt_host",
    "mqtt_port",
    "onboarding_endpoints",
    "onboarding_mode",
    "emitted_at",
)


def _is_non_empty_string(value: object) -> bool:
    return isinstance(value, str) and bool(value.strip())


def parse_bootstrap_payload(raw_payload: object) -> Tuple[bool, object]:
    try:
        if isinstance(raw_payload, bytes):
            text = raw_payload.decode("utf-8")
        elif isinstance(raw_payload, str):
            text = raw_payload
        else:
            text = str(raw_payload)
        return True, json.loads(text)
    except Exception:
        return False, "invalid_json"


def build_registration_url(api_base: str, register_path: str) -> str:
    api = str(api_base or "").strip()
    path = str(register_path or "").strip()
    if not _is_non_empty_string(api):
        raise ValueError("api_base is required")
    if not _is_non_empty_string(path):
        raise ValueError("onboarding_endpoints.register is required")

    parsed = urlparse(api)
    if not parsed.scheme or not parsed.netloc:
        raise ValueError("invalid api_base URL")

    base = f"{api.rstrip('/')}/"
    relative_path = path[1:] if path.startswith("/") else path
    base_path = parsed.path.strip("/")
    if base_path and (
        relative_path == base_path or relative_path.startswith(f"{base_path}/")
    ):
        relative_path = relative_path[len(base_path) :].lstrip("/")
    return urljoin(base, relative_path)


def validate_bootstrap_payload(
    payload: object,
    *,
    supported_versions: Iterable[int] = (1,),
    expected_topic: str = "synthia/bootstrap/core",
) -> Tuple[bool, object]:
    if not isinstance(payload, dict):
        return False, "invalid_payload"
    boundary_ok, boundary_error = enforce_bootstrap_security_boundary(payload)
    if not boundary_ok:
        return False, boundary_error
    if any(field not in payload for field in REQUIRED_FIELDS):
        return False, "missing_required_fields"

    register = payload.get("onboarding_endpoints", {}).get("register")
    if not _is_non_empty_string(register):
        return False, "missing_register_endpoint"

    topic = str(payload["topic"]).strip()
    bootstrap_version = int(payload["bootstrap_version"])
    core_id = str(payload["core_id"]).strip()
    core_name = str(payload["core_name"]).strip()
    core_version = str(payload["core_version"]).strip()
    api_base = str(payload["api_base"]).strip()
    mqtt_host = str(payload["mqtt_host"]).strip()
    mqtt_port = int(payload["mqtt_port"])
    onboarding_mode = str(payload["onboarding_mode"]).strip()
    emitted_at = str(payload["emitted_at"]).strip()
    register_endpoint = str(register).strip()

    if topic != expected_topic:
        return False, "invalid_topic"
    if bootstrap_version not in tuple(supported_versions):
        return False, "unsupported_bootstrap_version"
    if onboarding_mode != "api":
        return False, "unsupported_onboarding_mode"
    if not _is_non_empty_string(api_base):
        return False, "invalid_api_base"
    if not _is_non_empty_string(core_id) or not _is_non_empty_string(core_name):
        return False, "invalid_core_identity"
    if not _is_non_empty_string(core_version):
        return False, "invalid_core_version"
    if not _is_non_empty_string(mqtt_host) or mqtt_port <= 0:
        return False, "invalid_mqtt_target"
    if not _is_non_empty_string(emitted_at):
        return False, "invalid_emitted_at"

    try:
        registration_url = build_registration_url(api_base, register_endpoint)
    except ValueError:
        return False, "invalid_registration_url"

    return True, {
        "topic": topic,
        "bootstrap_version": bootstrap_version,
        "core_id": core_id,
        "core_name": core_name,
        "core_version": core_version,
        "api_base": api_base,
        "mqtt_host": mqtt_host,
        "mqtt_port": mqtt_port,
        "onboarding_endpoints": {"register": register_endpoint},
        "onboarding_mode": onboarding_mode,
        "emitted_at": emitted_at,
        "registration_url": registration_url,
    }

File: src/ai-node/bootstrap/bootstrapParser.js
const REQUIRED_FIELDS = [
  "topic",
  "bootstrap_version",
  "core_id",
  "core_name",
  "core_version",
  "api_base",
  "mqtt_host",
  "mqtt_port",
  "onboarding_endpoints",
  "onboarding_mode",
  "emitted_at",
];

function isNonEmptyString(value) {
  return typeof value === "string" && value.trim().length > 0;
}

function hasRequiredFields(payload) {
  return REQUIRED_FIELDS.every((key) => Object.hasOwn(payload, key));
}

function validateRegisterEndpoint(payload) {
  const register = payload.onboarding_endpoints?.register;
  return isNonEmptyString(register);
}

function normalizePayload(payload) {
  return {
    topic: payload.topic.trim(),
    bootstrap_version: Number(payload.bootstrap_version),
    core_id: payload.core_id.trim(),
    core_name: payload.core_name.trim(),
    core_version: payload.core_version.trim(),
    api_base: payload.api_base.trim(),
    mqtt_host: payload.mqtt_host.trim(),
    mqtt_port: Number(payload.mqtt_port),
    onboarding_endpoints: {
      register: payload.onboarding_endpoints.register.trim(),
    },
    onboarding_mode: payload.onboarding_mode.trim(),
    emitted_at: payload.emitted_at.trim(),
  };
}

export function buildRegistrationUrl(apiBase, registerPath) {
  const api = String(apiBase ?? "").trim();
  const path = String(registerPath ?? "").trim();

  if (!isNonEmptyString(api)) {
    throw new Error("api_base is required");
  }
  if (!isNonEmptyString(path)) {
    throw new Error("onboarding_endpoints.register is required");
  }

  const base = new URL(api.endsWith("/") ? api : `${api}/`);
  const pathWithoutLeadingSlash = path.startsWith("/") ? path.slice(1) : path;
  return new URL(pathWithoutLeadingSlash, base).toString();
}

export function parseBootstrapPayload(rawPayload) {
  try {
    const text =
      typeof rawPayload === "string"
        ? rawPayload
        : Buffer.isBuffer(rawPayload)
          ? rawPayload.toString("utf8")
          : String(rawPayload);

    return { ok: true, value: JSON.parse(text) };
  } catch {
    return { ok: false, error: "invalid_json" };
  }
}

export function validateBootstrapPayload(payload, options = {}) {
  const supportedVersions = options.supportedVersions ?? [1];
  const expectedTopic = options.expectedTopic ?? "synthia/bootstrap/core";

  if (!payload || typeof payload !== "object") {
    return { ok: false, error: "invalid_payload" };
  }
  if (!hasRequiredFields(payload)) {
    return { ok: false, error: "missing_required_fields" };
  }
  if (!validateRegisterEndpoint(payload)) {
    return { ok: false, error: "missing_register_endpoint" };
  }

  const normalized = normalizePayload(payload);
  if (normalized.topic !== expectedTopic) {
    return { ok: false, error: "invalid_topic" };
  }
  if (!supportedVersions.includes(normalized.bootstrap_version)) {
    return { ok: false, error: "unsupported_bootstrap_version" };
  }
  if (normalized.onboarding_mode !== "api") {
    return { ok: false, error: "unsupported_onboarding_mode" };
  }
  if (!isNonEmptyString(normalized.api_base)) {
    return { ok: false, error: "invalid_api_base" };
  }
  if (!isNonEmptyString(normalized.core_id) || !isNonEmptyString(normalized.core_name)) {
    return { ok: false, error: "invalid_core_identity" };
  }
  if (!isNonEmptyString(normalized.mqtt_host) || !Number.isFinite(normalized.mqtt_port)) {
    return { ok: false, error: "invalid_mqtt_target" };
  }
  if (!isNonEmptyString(normalized.emitted_at)) {
    return { ok: false, error: "invalid_emitted_at" };
  }

  try {
    return {
      ok: true,
      value: {
        ...normalized,
        registration_url: buildRegistrationUrl(
          normalized.api_base,
          normalized.onboarding_endpoints.register,
        ),
      },
    };
  } catch {
    return { ok: false, error: "invalid_registration_url" };
  }
}
